### PR TITLE
Anya/5053 judge assign to another attorney

### LINKED ISSUE #5053: Allow judges to change an appeal assignment from one attorney to another
We need to allow judges to change an appeal assignment from one attorney to another. 

## Open Questions
- Should we use `PATCH /tasks/:task_id`? 
  - `deassign` is currently a component of the task id. And that will change when a case is reassigned. So effectively once we `PATCH`, the task id will change
- Should we do a combination of `DELETE /tasks/:task_id` and `POST /tasks`? 
  - How do we make it into a transaction? It's not acceptable for one to fail and the other to succeed.
- Should we delete a DECASS row, or just change the attributes of the existing DECASS row?

### PR DESCRIPTION
**Pull request by aroltsch**
Resolves #5053 

### Description
Allow a judge to assign a case to another attorney

### Testing Plan
Through postman

```
curl -X PATCH 
  http://localhost:3000/queue/tasks/3615398-2018-04-18 
  -d '{ "appeal_id": 103,
  "appeal_type": "Legacy",
  "attorney_id": 1533
}'
```

Through command line:
```
params = { appeal_id: 103, appeal_type: "Legacy", assigned_to: User.find(1533), assigned_by: User.find(1278), task_id: "3615398-2018-04-18" }
JudgeCaseAssignment.new(params).reassign_to_attorney!
```

### CODE REVIEW

File: app/controllers/queue_controller.rb [hunk 1]
```diff
@@ -31,14 +30,17 @@ def complete
   end
 
   def create
-    return invalid_appeal_type unless APPEAL_TYPES.include?(create_params[:appeal_type])
```
Comment (aroltsch) [inline app/controllers/queue_controller.rb]: Moved this validation to JudgeCaseAssignment.rb

File: app/mappers/queue_mapper.rb [hunk 2]
```diff
@@ -30,11 +34,8 @@ def self.rename_and_validate_decass_attrs(decass_attrs)
       result[COLUMN_NAMES[k]] = decass_attrs[k]
       result
     end
-    VacolsHelper.validate_presence(update_attrs, [:deprod, :dedocid])
```
Comment (aroltsch) [inline app/mappers/queue_mapper.rb]: Moved this validation to JudgeCaseAssignment.rb

File: app/controllers/queue_controller.rb [hunk 3]
```diff
@@ -31,14 +30,17 @@ def complete
...
-
     return invalid_role_error if current_user.vacols_role != "Judge"
-
-    JudgeCaseAssignment.assign_to_attorney!(create_params)
+    JudgeCaseAssignment.new(task_params).assign_to_attorney!
     render json: {}, status: :created
   end
 
+  def update
+    return invalid_role_error if current_user.vacols_role != "Judge"
```
Comment (amprokop) [inline app/controllers/queue_controller.rb]: 👍

**Comment by amprokop** [comment]
## Testing
```
appeal_id = 43
appeal_type = "Legacy"
assigned_to =  User.find_by(css_id: "BVAJSCHIMMEL")
assigned_by =  User.find_by(full_name: "Anjali Q Abshire")
task_id = "3019694-2017-12-06"
RequestStore[:current_user] = assigned_by

params = { appeal_id: appeal_id, appeal_type: appeal_type, assigned_to: assigned_to, assigned_by: assigned_by, task_id: task_id}

JudgeCaseAssignment.new(params).reassign_to_attorney!
```

Result:

UI showed that case was successfully reassigned from BVASRITCHIE to BVAJSCHIMMEL

File: app/models/judge_case_assignment.rb [hunk 4]
```diff
@@ -1,22 +1,57 @@
...
+  validates :task_id, format: { with: /\A[0-9]+-[0-9]{4}-[0-9]{2}-[0-9]{2}\Z/i }, allow_blank: true
+  validates :appeal_type, inclusion: { in: %w[Legacy Ama] }
+  validates :assigned_by, :assigned_to, presence: true
+
+  def assign_to_attorney!
+    validate!
+    case appeal_type
+    when "Legacy"
+      vacols_id = Appeal.find(appeal_id).vacols_id
+      MetricsService.record("VACOLS: assign_case_to_attorney #{vacols_id}",
```
Comment (amprokop) [inline app/models/judge_case_assignment.rb]: Can we track this as a business metric too?

File: app/models/judge_case_assignment.rb [hunk 5]
```diff
@@ -1,22 +1,57 @@
...
-        MetricsService.record("VACOLS: assign_case_to_attorney #{vacols_id}",
-                              service: :vacols,
-                              name: "assign_case_to_attorney") do
-          repository.assign_case_to_attorney!(
-            judge: params[:assigned_by],
-            attorney: params[:assigned_to],
-            vacols_id: vacols_id
-          )
-        end
+  def reassign_to_attorney!
```
Comment (amprokop) [inline app/models/judge_case_assignment.rb]: Can we track assignment/reassignment as business metrics too?

Comment (aroltsch) [reply app/models/judge_case_assignment.rb]: So MetricsService.record does two things: records latency and counting attempts. I think BusinessMetrics needs to be used when we don't care about latency. In this case we care about latency and count.

Comment (amprokop) [reply app/models/judge_case_assignment.rb]: Got it

**Pull request merged**